File: frontend/src/lib/providers/rakuten_bank.ts
import type { AccountCredentials, AccountProvider, CredentialField, SyncResult } from "./index";

export class RakutenBankProvider implements AccountProvider {
  providerCode = "rakuten_bank";
  providerName = "楽天銀行";

  requiredFields(): CredentialField[] {
    return [
      { key: "login_id", label: "ユーザーID", field_type: "text", required: true },
      { key: "password", label: "ログインパスワード", field_type: "password", required: true },
    ];
  }

  validateCredentials(creds: AccountCredentials): void {
    if (!creds.login_id) throw new Error("ユーザーIDを入力してください");
    if (!creds.password) throw new Error("パスワードを入力してください");
  }

  async sync(_accountId: string, _credentials: AccountCredentials): Promise<SyncResult> {
    throw new Error("楽天銀行の自動取得は実装中です。");
  }
}

File: backend/src/providers/rakuten_bank.rs
use anyhow::{anyhow, Result};
use async_trait::async_trait;
use uuid::Uuid;

use super::{AccountCredentials, AccountProvider, CredentialField, SyncResult};

pub struct RakutenBankProvider;

#[async_trait]
impl AccountProvider for RakutenBankProvider {
    fn provider_code(&self) -> &'static str {
        "rakuten_bank"
    }

    fn provider_name(&self) -> &'static str {
        "楽天銀行"
    }

    fn required_fields(&self) -> Vec<CredentialField> {
        vec![
            CredentialField {
                key: "login_id".to_string(),
                label: "ユーザーID".to_string(),
                field_type: "text".to_string(),
                required: true,
                placeholder: Some("楽天銀行のユーザーID".to_string()),
                help_text: None,
            },
            CredentialField {
                key: "password".to_string(),
                label: "ログインパスワード".to_string(),
                field_type: "password".to_string(),
                required: true,
                placeholder: None,
                help_text: None,
            },
        ]
    }

    fn validate_credentials(&self, creds: &AccountCredentials) -> Result<()> {
        if creds.login_id.as_deref().unwrap_or("").is_empty() {
            return Err(anyhow!("ユーザーIDを入力してください"));
        }
        if creds.password.as_deref().unwrap_or("").is_empty() {
            return Err(anyhow!("パスワードを入力してください"));
        }
        Ok(())
    }

    async fn sync(&self, account_id: Uuid, _credentials: &AccountCredentials) -> Result<SyncResult> {
        tracing::info!("RakutenBankProvider::sync called for account {}", account_id);
        Err(anyhow!("楽天銀行の自動取得は実装中です。"))
    }
}
